core/arch/aarch64/emit_utils.c

```diff
@@ -184,33 +273,63 @@ patch_stub(fragment_t *f, cache_pc stub_pc, cache_pc target_pc, bool hot_patch)
     /* Compute offset as unsigned, modulo arithmetic. */
     ptr_uint_t off = (ptr_uint_t)target_pc - (ptr_uint_t)stub_pc;
     if (off + 0x8000000 < 0x10000000) {
-        /* We can get there with a B (OP_b, 26-bit signed immediate offset). */
-        *(uint *)vmcode_get_writable_addr(stub_pc) =
-            (0x14000000 | (0x03ffffff & off >> 2));
+        /* target_pc is a near fragment. We can get there with a B
+         * (OP_b, 26-bit signed immediate offset).
+         * i#1911: Patching arbitrary instructions to an unconditional branch
+         * is theoretically not sound. Architectural specifications do not
+         * guarantee safe behaviour or any bound on when the change will be
+         * visible to other process elements.
+         */
+        uint b_enc = 0x14000000 | (0x03ffffff & off >> 2);
+        ATOMIC_4BYTE_ALIGNED_WRITE(vmcode_get_writable_addr(stub_pc), b_enc);
         if (hot_patch)
             machine_cache_sync(stub_pc, stub_pc + 4, true);
         return;
     }
-    /* We must use an indirect branch. */
-    ASSERT_NOT_IMPLEMENTED(false); /* FIXME i#1569 */
+    /* target_pc is a far fragment. We must use an indirect branch. */
+    byte *target_pc_slot = get_target_pc_slot(f, stub_pc);
+    /* We set hot_patch to false as we are not modifying code. */
+    ATOMIC_8BYTE_ALIGNED_WRITE(target_pc_slot, (ptr_uint_t)target_pc,
+                               /*hot_patch=*/false);
+    return;
 }
 
 bool
-stub_is_patched(fragment_t *f, cache_pc stub_pc)
+stub_is_patched(dcontext_t *dcontext, fragment_t *f, cache_pc stub_pc)
 {
-    ASSERT_NOT_IMPLEMENTED(false); /* FIXME i#1569 */
-    return false;
+    uint enc;
+    ATOMIC_4BYTE_ALIGNED_READ(vmcode_get_writable_addr(stub_pc), &enc);
+    /* linked to intermediate fragment. */
+    if ((enc & 0x14000000) == 0x14000000)
+        return true;
+
+    ptr_uint_t target_pc;
+    ATOMIC_8BYTE_ALIGNED_READ(get_target_pc_slot(f, stub_pc), &target_pc);
+    /* linked to far fragment. */
+    return target_pc != (ptr_uint_t)fcache_return_routine(dcontext);
 }
 
 void
-unpatch_stub(fragment_t *f, cache_pc stub_pc, bool hot_patch)
+unpatch_stub(dcontext_t *dcontext, fragment_t *f, cache_pc stub_pc, bool hot_patch)
```

Comment:
This implementation consists of two operations to unpatch, but at any given time only one patching strategy would be used. So, one of these two operations would really be a no-op.

Reply:
Resolved in other comment thread.